File: typescript/packages/core/src/commands/builtin/ram/rg.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { readdir as ramReaddir } from '../../../core/ram/readdir.ts'
import { stat as ramStat } from '../../../core/ram/stat.ts'
import { stream as ramStream } from '../../../core/ram/stream.ts'
import { find as ramFind } from '../../../core/ram/find.ts'
import type { RAMAccessor } from '../../../accessor/ram.ts'
import { ResourceName } from '../../../types.ts'
import { command } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { rgGeneric } from '../generic/rg.ts'

export const RAM_RG = command({
  name: 'rg',
  resource: ResourceName.RAM,
  spec: specOf('rg'),
  fn: (accessor: RAMAccessor, paths, texts, opts) =>
    rgGeneric(
      paths,
      texts,
      opts,
      (p) => ramStat(accessor, p),
      (p) => ramReaddir(accessor, p),
      (p) => ramStream(accessor, p),
      (root, options) => ramFind(accessor, root, options),
    ),
})

File: python/mirage/commands/builtin/discord/rg.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import logging
from collections.abc import AsyncIterator

from mirage.accessor.discord import DiscordAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.generic.rg import rg as generic_rg
from mirage.commands.builtin.utils.output import format_records
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.discord.channels import list_channels
from mirage.core.discord.entry import channel_dirname
from mirage.core.discord.formatters import format_grep_results
from mirage.core.discord.glob import resolve_glob
from mirage.core.discord.read import read as discord_read
from mirage.core.discord.readdir import readdir as _readdir
from mirage.core.discord.scope import coalesce_scopes, detect_scope
from mirage.core.discord.search import search_guild
from mirage.core.discord.stat import stat as _stat
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec

logger = logging.getLogger(__name__)


@command("rg", resource="discord", spec=SPECS["rg"])
async def rg(
    accessor: DiscordAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    prefix: str = "",
    index: IndexCacheStore = None,
    **flags: object,
) -> tuple[ByteSource | None, IOResult]:
    e = flags.get("e")
    if not isinstance(e, str) and not texts:
        raise ValueError("rg: usage: rg [flags] pattern [path]")
    pattern_str = e if isinstance(e, str) else texts[0]
    m = flags.get("m")
    max_count = int(m) if isinstance(m, str) else None

    pushdown_warnings: list[str] = []
    if paths:
        scope = await detect_scope(paths[0], index)
        if scope.level in ("messages", "file_blob", "date"):
            coalesced = await coalesce_scopes(paths, index)
            if coalesced is not None:
                scope = coalesced

        if scope.level == "root":
            return b"", IOResult(exit_code=1,
                                 stderr=b"rg: root-level search "
                                 b"not yet supported\n")

        if scope.level in ("channel", "guild"):
            try:
                if scope.guild_id is None:
                    raise RuntimeError("cannot resolve guild ID")
                msgs = await search_guild(
                    accessor.config,
                    scope.guild_id,
                    pattern_str,
                    channel_id=scope.channel_id,
                    limit=max_count or 100,
                )
                file_prefix = paths[0].prefix or ""
                resource_first = scope.resource_path.split("/", 1)[0]
                channels = await list_channels(accessor.config, scope.guild_id)
                channel_map = {c["id"]: channel_dirname(c) for c in channels}
                lines = format_grep_results(msgs, file_prefix, resource_first,
                                            channel_map)
                if not lines:
                    return b"", IOResult(exit_code=1)
                return format_records(lines), IOResult()
            except Exception as exc:
                msg = str(exc)
                pushdown_warnings.append(
                    f"discord: native search push-down failed ({msg}); "
                    f"falling back to per-file scan")
                if ("403" in msg or "Forbidden" in msg
                        or "missing access" in msg.lower()):
                    pushdown_warnings.append(
                        "discord: hint - ensure the bot has the "
                        "READ_MESSAGE_HISTORY permission for this guild "
                        "and the MESSAGE CONTENT privileged intent enabled")
                logger.warning(
                    "discord search push-down failed (%s); "
                    "falling back to per-file scan", exc)

        paths = await resolve_glob(accessor, paths, index=index)

    stdout, io = await generic_rg(
        paths,
        texts,
        flags,
        readdir=_readdir,
        stat=_stat,
        read_bytes=discord_read,
        read_stream=None,
        accessor=accessor,
        stdin=stdin,
        index=index,
    )
    if pushdown_warnings:
        io.stderr = ("\n".join(pushdown_warnings) + "\n").encode()
    return stdout, io
